server/server.go

```diff
@@ -32,55 +62,170 @@ func StartHttpServer(opts *options.Options) error {
+func checkHealthChecks(opts *options.Options) *httpResponse {
 	logger := opts.Logger
 	logger.Infof("Received inbound request. Beginning health checks...")
 
-	allPortsValid := true
-
+	// initialize failedChecks to 0, used as atomic counter for goroutines below
+	var failedChecks uint64
 	var waitGroup = sync.WaitGroup{}
 
-	for _, port := range opts.Ports {
+	for _, check := range opts.Checks {
 		waitGroup.Add(1)
-		go func(port int) {
-			err := attemptTcpConnection(port, opts)
+		go func(check options.Check) {
+			defer waitGroup.Done()
+			err := check.DoCheck(opts)
 			if err != nil {
-				logger.Warnf("TCP connection to port %d FAILED: %s", port, err)
-				allPortsValid = false
-			} else {
-				logger.Infof("TCP connection to port %d successful", port)
+				logger.Warnf(err.Error())
+				atomic.AddUint64(&failedChecks, 1)
 			}
-
-			waitGroup.Done()
-		}(port)
+		}(check)
 	}
 
 	waitGroup.Wait()
 
-	if allPortsValid {
-		logger.Infof("All health checks passed. Returning HTTP 200 response.\n")
-		return &httpResponse{ StatusCode: http.StatusOK, Body: "OK" }
-	} else {
+	failedChecksFinal := atomic.LoadUint64(&failedChecks)
+	if failedChecksFinal > 0 {
 		logger.Infof("At least one health check failed. Returning HTTP 504 response.\n")
-		return &httpResponse{ StatusCode: http.StatusGatewayTimeout, Body: "At least one health check failed" }
+		return &httpResponse{StatusCode: http.StatusGatewayTimeout, Body: "At least one health check failed"}
+	} else {
+		logger.Infof("All health checks passed. Returning HTTP 200 response.\n")
+		return &httpResponse{StatusCode: http.StatusOK, Body: "OK"}
+	}
+}
+
+func (c TcpCheck) ValidateCheck () error {
+	if c.Name == "" {
+		return &InvalidCheck{name: "tcp", key: "name"}
+	}
+	if c.Host == "" {
+		return &InvalidCheck{name: c.Name, key: "host"}
 	}
+	if c.Port == 0 {
+		return &InvalidCheck{name: c.Name, key: "port"}
+	}
+	return nil
 }
 
-// Attempt to open a TCP connection to the given port
-func attemptTcpConnection(port int, opts *options.Options) error {
+func (c TcpCheck) DoCheck (opts *options.Options) error {
 	logger := opts.Logger
-	logger.Infof("Attempting to connect to port %d via TCP...", port)
+	logger.Infof("Checking %s at %s:%d via TCP...", c.Name, c.Host, c.Port)
 
-	defaultTimeout := time.Second * 5
+	timeout := time.Second * DEFAULT_CHECK_TIMEOUT
+	if c.Timeout != 0 {
+		// override default with user defined timeout
+		timeout = time.Second * time.Duration(c.Timeout)
+	}
 
-	conn, err := net.DialTimeout("tcp", fmt.Sprintf("0.0.0.0:%d", port), defaultTimeout)
+	conn, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%d", c.Host, c.Port), timeout)
 	if err != nil {
-		return err
+		return &CheckFail{name: c.Name, reason: err.Error()}
 	}
 
 	defer conn.Close()
+	logger.Infof("Check SUCCESS: %s", c.Name)
+	return nil
+}
 
+func (c HttpCheck) ValidateCheck () error {
+	if c.Name == "" {
+		return &InvalidCheck{name: "http", key: "name"}
```

Comment:
Why return an `InvalidCheck` pointer in all of these? Why not just a normal value?